### PR TITLE
feat(v2): allow adding components to react-live scope

### PR DESCRIPTION
**Pull request by slorber**
## Motivation

Currently, the react-live scope only exposes the React exports, so we can use React, useState etc, but we can't import any other code easily.

If you build a react component library, and want to allow users to play with the components in a live editor, it's important to have a documented way to provide those components in the react-live scope.

See https://github.com/facebook/docusaurus/issues/2807

## Test Plan

Dogfooding by using this feature on the Docusaurus website directly

## Notes

Swizzling of the react-live scope object is not an ideal solution, as it's global and needs more work from the user, but I think it's good enough to solve the usecase for now.

When we'll have a way to wrap root component, users could use the MDXProvider to reuse the components of the MDX scope for react-live too, instead of swizzling. 

Gatsby has a cool feature that permits to make available in react-live scope anything that is imported in the MDX file. Unfortunately, this feature is not so simple to port to Docusaurus (involves extracting imports in mdx string using a babel plugin, writing them to disk etc... see https://github.com/facebook/docusaurus/issues/2807#issuecomment-635268531)

### CODE REVIEW

File: packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js [hunk 1]
```diff
@@ -9,18 +9,33 @@ import React from 'react';
...
+import ReactLiveScope from '@theme/ReactLiveScope';
 import CodeBlock from '@theme-init/CodeBlock';
+import {useMDXComponents} from '@mdx-js/react';
+
+// Returns components that are available to use by the react-live playground
+// It's not possible to import anything in the live playground,
+// we need to provide the available imports manually
+// See https://github.com/facebook/docusaurus/issues/2807
+const useReactLiveScope = () => {
+  const mdxComponents = useMDXComponents();
```
Comment (slorber) [inline packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js:21]: not 100% sure this is needed as of today to provide the MDX components, as afaik we don't have a way to use the MDXProvider for now, but in the future (with wrapRootElement api) that might be a good default behavior

**Comment by slorber** [comment]
Direct link to test: https://deploy-preview-2826--docusaurus-2.netlify.app/docs/next/markdown-features#interactive-code-editor

image

File: packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js (hunk 1 above)
Comment (lex111) [reply packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js:21]: Maybe we should _not_ include the built-in MDX components (`a`, `code`, `h1` etc). Does this bring any overhead?

Comment (slorber) [reply packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js:21]: Actually, I tested and using in react-live `<a>` does not even pick the `a` declared in the scope, but the regular ReactDOM `a`. 

I guess it's due to their Bublé code transpilation. 

Didn't check, but it's possible that to use the `a` component available in scope, you have to do something like `React.createElement(a,props)`


If we are not so sure about providing the `useMDXComponents()` components in react-live scope, I can remove it from now and we'll see later to introduce it back if it makes sense.

Comment (lex111) [reply packages/docusaurus-theme-live-codeblock/src/theme/CodeBlock/index.js:21]: Let's better remove it, and only leave the ability to add custom components so as not to complicate this PR.

**Comment by slorber** [comment]
Just removed the mdx comps from react-live scope @lex111

File: packages/docusaurus-theme-live-codeblock/package.json [hunk 2]
```diff
@@ -9,6 +9,7 @@
   "license": "MIT",
   "dependencies": {
     "@philpl/buble": "^0.19.7",
+    "@mdx-js/react": "^1.5.8",
```
[the remark this answers is not in the document]
Comment (slorber) [reply packages/docusaurus-theme-live-codeblock/package.json]: done

**Comment by lex111** [comment]
@slorber thanks!

**Comment by wintercounter** [comment]
Is this released already? I'm getting this error:

```
Component Ui was not imported, exported, or provided by MDXProvider as global scope

ReferenceError: Ui is not defined
(undefined) ReferenceError: Ui is not defined
```

There is a component defined as `<Ui>`. Previously I was simply doing `Object.assign(React, { ...myProps })` using a custom plugin, but it stopped working. Then I found this PR, it's still not working with it either.

```
    "@docusaurus/core": "^2.0.0-alpha.61",
    "@docusaurus/preset-classic": "^2.0.0-alpha.61",
    "@docusaurus/theme-live-codeblock": "^2.0.0-alpha.61",
```

**Comment by slorber** [comment]
@wintercounter I can't tell anything if you don't show anycode, but did you swizzle `@theme/ReactLiveScope` as documented?

```
const ReactLiveScope = {
  React,
  ...React,
};
```

We spread `...React` at initialization time, so when this is done, you might not have assigned your Ui comp to React yet, so it is not in the scope anymore.

I don't think assigning Ui to React is a good idea, you'd rather register the comp explicitly in `ReactLiveScope`

**Comment by wintercounter** [comment]
Turns out my file was cached somewhere and it didn't update, seems like it's working now, however, I still get the message.

```
> docusaurus build

Creating an optimized production build...

√ Client
  Compiled successfully in 47.77s

√ Server
  Compiled successfully in 36.40s

Component Ui was not imported, exported, or provided by MDXProvider as global scope

Success! Generated static files in build.Use `npm run serve` to test your build locally.
```

**Pull request merged**